### PR TITLE
Make blended backend a real backend

### PR DESCRIPTION
**Pull request by peterbarker**
Tested in SITL.

```
Board               AP_Periph  blimp  bootloader  copter  heli  iofirmware  plane  rover  sub
Durandal                       352    *           184     304               328    352    360
Hitec-Airspeed      *                                                                     
KakuteH7-bdshot                480    *           496     472               656    472    472
MatekF405                      56     *           128     128               -96    80     104
Pixhawk1-1M-bdshot             96                 120     104               64     104    104
f103-QiotekPeriph   *                                                                     
f303-Universal      *                                                                     
iomcu                                                           *                         
revo-mini                      64     *           -96     8                 16     0      104
skyviper-v2450                                    *                                       
```

### CODE REVIEW

File: libraries/AP_GPS/AP_GPS_Blended.h [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    // GPS blending and switching
+    Vector3f _blended_antenna_offset; // blended antenna offset
+    float _blended_lag_sec; // blended receiver lag in seconds
+    float _blend_weights[GPS_MAX_RECEIVERS]; // blend weight for each GPS. The blend weights must sum to 1.0 across all instances.
+    bool _output_is_blended; // true when a blended GPS solution being output
+    uint8_t _blend_health_counter;  // 0 = perfectly health, 100 = very unhealthy
+
+    // calculate the blend weight.  Returns true if blend could be
+    // calculated, false if not
+    // bool calc_weights(void);
```
Comment (amilcarlucas) [inline libraries/AP_GPS/AP_GPS_Blended.h:69]: Remove this, and move the comments to the public side

File: libraries/AP_GPS/AP_GPS_Blended.h [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+    // calculate the blended state
```
Comment (amilcarlucas) [inline libraries/AP_GPS/AP_GPS_Blended.h:71]: Remove this and move the comments to the top of the function declaration on the public side.

**Review by amilcarlucas — CHANGES_REQUESTED** [review]
Very cool changes!

File: libraries/AP_GPS/AP_GPS_Blended.h (hunk 1 above)
Comment (peterbarker) [reply libraries/AP_GPS/AP_GPS_Blended.h]: Done

File: libraries/AP_GPS/AP_GPS_Blended.h (hunk 2 above)
Comment (peterbarker) [reply libraries/AP_GPS/AP_GPS_Blended.h]: Done

**Comment by tridge** [comment]
we will do some SpringValley flight testing

**Comment by peterbarker** [comment]
Flight testing at SV went well today.

**Review by tridge — CHANGES_REQUESTED** [review]
how does this interact with EKF3 affinity?

**Comment by tridge** [comment]
I'd like to understand how this interacts with affinity

File: libraries/AP_GPS/AP_GPS.h [hunk 3]
```diff
@@ -606,7 +604,7 @@ class AP_GPS
 protected:
 
     // configuration parameters
-    Params params[GPS_MAX_RECEIVERS];
+    Params params[GPS_MAX_INSTANCES];
```
[the remark this answers is not in the document]
Comment (peterbarker) [reply libraries/AP_GPS/AP_GPS.h:608]: Not yet.  There are all sorts of interesting decisions to be made revolving around it.

Many of them will end up turning into a `:`-style iteration across the instances array.

**Comment by amilcarlucas** [comment]
Is this planned for 4.6?

**Comment by peterbarker** [comment]
Out GPS simulation's a bit weird (this is master):
image

**Comment by peterbarker** [comment]
I've flown a real copter with blended a bunch now - even with different GPS types it copes.

I've added an autotest that tests blended/affinity before and after.

The logs on master show that when affinity is in play the third core will use the blended GPS.

**Comment by Ryanf55** [comment]
I tested this in SITL a while back using some of the SIM* params to muck with the GPS health of GPS 1 and 2 and could not get the vehicle to have any serious issues with blended. I was wanting triple-GPS support, and was interested in this.
I haven't tested the latest version.

The new test is nice to ensure GPS errors get blended out.

**Review by tridge — APPROVED** [review]

**Pull request merged**